commit b02e5ed986bbeb45e0f82b0a6f71104ccf5a93b9
Author: Li Guanglei <guanglei.li@unisoc.com>
Date:   Tue Oct 27 10:55:58 2020 +0800

    eas: eas_big_to_small: fix wrong big{small}_task_us in corner case
    
    sometimes the task running as big was scheduled after "SMALL TASK"
    trace marker:
    
    eas_big_to_smal [big core] .... 27284.761959: SMALL TASK
    eas_big_to_smal [big core] d..2 27284.761978: sched_switch:
      prev_comm=eas_big_to_small ==> next_comm=swapper/6
    
    the last execution segment of big task is counted as small_task_us
    in this case.
    
    after "SMALL TASK" trace marker, the small task will be scheduled
    for the first time, mark this time as smalltask_ts_usec can fix
    this issue.
    
    Bug: 172186049
    Signed-off-by: Li Guanglei <guanglei.li@unisoc.com>
    Change-Id: I073b7a58a5e1188fe60f3c623e74cc7b1b4ca73b

diff --git a/testcases/kernel/sched/eas/eas_big_to_small.c b/testcases/kernel/sched/eas/eas_big_to_small.c
index 36f9598cd..8af883699 100644
--- a/testcases/kernel/sched/eas/eas_big_to_small.c
+++ b/testcases/kernel/sched/eas/eas_big_to_small.c
@@ -50,6 +50,7 @@ static int parse_results(void)
 	unsigned long long big_task_us = 0;
 	unsigned long long small_task_us = 0;
 	unsigned long long smalltask_ts_usec = 0;
+	unsigned long long smalltask_tm_usec = 0;
 	unsigned long long downmigrate_ts_usec = 0;
 	unsigned long long downmigrate_latency_usec = 0;
 	cpu_set_t cpuset;
@@ -62,10 +63,11 @@ static int parse_results(void)
 	for (i = 0; i < num_trace_records; i++) {
 		unsigned long long segment_us;
 		struct trace_sched_switch *t = trace[i].event_data;
+		unsigned long long trace_ts_usec = TS_TO_USEC(trace[i].ts);
 
 		if (trace[i].event_type == TRACE_RECORD_TRACING_MARK_WRITE &&
 			   !strcmp(trace[i].event_data, "SMALL TASK")) {
-			smalltask_ts_usec = TS_TO_USEC(trace[i].ts);
+			smalltask_tm_usec = trace_ts_usec;
 			continue;
 		}
 
@@ -73,12 +75,15 @@ static int parse_results(void)
 			continue;
 
 		if (t->next_pid == task_tid) {
+			if (!smalltask_ts_usec && smalltask_tm_usec &&
+					trace_ts_usec > smalltask_tm_usec)
+				smalltask_ts_usec = trace_ts_usec;
 			/* Start of task execution segment. */
 			if (exec_start_us) {
 				printf("Trace parse fail: double exec start\n");
 				return -1;
 			}
-			exec_start_us = TS_TO_USEC(trace[i].ts);
+			exec_start_us = trace_ts_usec;
 			if (smalltask_ts_usec && !downmigrate_ts_usec &&
 			    CPU_ISSET(trace[i].cpu, &cpuset))
 				downmigrate_ts_usec = exec_start_us;
@@ -87,7 +92,7 @@ static int parse_results(void)
 		if (t->prev_pid != task_tid)
 			continue;
 		/* End of task execution segment. */
-		segment_us = TS_TO_USEC(trace[i].ts);
+		segment_us = trace_ts_usec;
 		segment_us -= exec_start_us;
 		exec_start_us = 0;
 		if (CPU_ISSET(trace[i].cpu, &cpuset)) {
